### PR TITLE
Added recommendation system

### PR DESCRIPTION
**Pull request by rahul3613**
Added a section for recommendations on the home page. Which will recommend files to the users based on their similarity with other users in accessing files.

### CODE REVIEW

**Comment by paniash** [comment]
Hi! Could you correct the PEP styling as the bot mentioned?

File: main/templates/main/recom.htm
Comment (JeS24) [inline main/templates/main/recom.htm:70]: `{% if auth %}` works.

File: main/templates/main/recom.htm
Comment (JeS24) [inline main/templates/main/recom.htm:76]: Styles in .css as much as possible.

File: main/recom.py
Comment (JeS24) [inline main/recom.py:16]: Rename this function to something more informational, like get_recs() or get_recommendations() etc.

File: main/views.py
Comment (JeS24) [inline main/views.py:24]: Use a separate python script (file) to start this and keep this running.

File: requirements.txt
Comment (JeS24) [inline requirements.txt:6]: `lockfile` is deprecated and not under development since 2015. Perhaps you can use [fasteners](https://pypi.org/project/fasteners/).

File: main/recom.py
Comment (JeS24) [inline main/recom.py:33]: If the possibilities that come out from the `if` statement are mutually exclusive, then you don't need the `else` statement. For example,
```
if 2 > 0:
    return 2
else:
    return 0
```
can be written as:
```
if 2 > 0:
    return 2

return 0
```
In your code, you could write:
```
if user_id in df.columns:
    if len(df)>0:
        corr = df.corr()[user_id]
        ...
        return final.index.tolist()   

    return []


while lock.is_locked():
    sleep(0.05)

...

return []
```

This applies to a lot of other places in your code. But do check the logic to ensure that mutual exclusivity is present.

**Review by JeS24 — CHANGES_REQUESTED** [review]
Short review of the code. Additionally, you need to do the following:
1. Change function names e.g. update1 and file_view1. Make them slightly more descriptive. Or, try to merge them with the original functions, update and file_view.
2. Put comments and docstrings in all files and functions and methods that you have added.
3. Explain how the recommendation system works. You could do it through dosctrings in the code or you can write a [Gist](https://gist.github.com/), that we can later add as a Wiki page for this repo. It will be helpful for repo maintainers and future contributors.
4. After all the code (+ comments) is finalized, run [`black`](https://pypi.org/project/black/) on the code to properly format it.

Finally, we will need to squash the commits before merging.

**Comment by rahul3613** [comment]
Hi, @JeS24 I have made all the changes suggested by you. Please have a look,
Here is the [link for the Gist](https://gist.github.com/rahul3613/f0c4e4bbf5377c2c933f9d077f8b001f) explaining how the recommendation system works. (I don't know where to put it)
Regarding 'squashing the commits' I think it can be easily done while merging from your side?

**Comment by 11DE784A** [comment]
Hi. Thanks for making the changes. Here are a few suggestions:

1. Make the recommendations section on the homepage less intrusive, just add a single line which has a list of recommendations. Instead of
Screenshot 2022-02-24 at 18-09-08 Home NISER Archive
have something like
Screenshot 2022-02-24 at 18-52-42 Home NISER Archive

2. Instead of calling `read_csv` everytime with `update` and `get_recom` declare a global dataframe and save the variable to file occasionally.

File: main/recom.py
Comment (11DE784A) [inline main/recom.py:103]: This statement occurs in each of the four conditions. You should move this to outside the entire if block.

File: main/templates/main/recom.htm
Comment (11DE784A) [inline main/templates/main/recom.htm:8]: As before, move style information to CSS

File: main/views.py
Comment (11DE784A) [inline main/views.py:271]: Writing two functions with almost identical contents is really bad.

File: recom_data.csv
Comment (11DE784A) [inline recom_data.csv:1]: No need to track this file. Add to gitignore.

**Comment by 11DE784A** [comment]
Everything looks fine to me. Will merge soon.

**Comment by JeS24** [comment]
@rahul3613 The description in the gist is nicely presented. We are almost ready to merge. One last change, I would need from your side is if you can use either the numpydoc or the Google documentation system, since those are standard and can make parsing very simple for existing docs-related packages. And, run `black` on the files you have modified, so that, we conform to PEP8 norms.

@11DE784A We can merge, if no runtime issues have come up in your local testing. Please do squash the commits.

**Comment by rahul3613** [comment]
Hi, @JeS24 & @11DE784A I have changed the docstring format to "numpydoc". Please have a look.

**Comment by 11DE784A** [comment]
Looks good to me.
Please merge and close @Abhishek-Deshmukh

**Comment by JeS24** [comment]
Looks good. We can merge.

**Comment by JeS24** [comment]
Thanks @rahul3613.

**Pull request merged**
